src/middleware/timeout/index.ts

```diff
@@ -0,0 +1,61 @@
+import { HTTPException } from '../../http-exception'
+import type { MiddlewareHandler } from '../../types'
+import type { StatusCode } from '../../utils/http-status'
+
+interface TimeoutOptions {
+  message?: string
+  code?: StatusCode
+}
+
+const DEFAULT_ERROR_MESSAGE = 'Gateway Timeout'
+const DEFAULT_ERROR_CODE = 504
+
+interface DurationUnits {
+  [key: string]: number
+}
+
+const parseDuration = (duration: string): number => {
+  const units: DurationUnits = { ms: 1, s: 1000, m: 60000, h: 3600000 }
+  const pattern = /(\d+)(ms|s|m|h)/g
+  let totalMilliseconds = 0
+
+  let match: RegExpExecArray | null
+  while ((match = pattern.exec(duration)) !== null) {
+    const value = parseInt(match[1], 10)
+    const unit = match[2]
+
+    if (!units[unit]) {
+      throw new Error(`Unsupported time unit: ${unit}`)
+    }
+    totalMilliseconds += value * units[unit]
+  }
+
+  return totalMilliseconds
+}
+
+export const timeout = (
+  duration: number | string,
+  options: TimeoutOptions = {}
```

Comment:
How about taking `exception` as the second arg:

```ts
export const timeout = (
  duration: number | string,
  exception?: HTTPException | ((c: Context) => HTTPException)
)
```

How to use it (callback pattern):

```ts
app.get(
  '/',
  timeout(1000, (c) => {
    return new HTTPException(408, {
      message: `Timeout occurred at ${c.req.path}`,
    })
  }),
  async (c) => {
    // ...
  }
)
```

Reply:
Nice. I fixed it, can you see if the type definitions and other perceptions you think match?